File: partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        
        targetSum = sum(nums)
        
        if targetSum % 2 != 0:
            return False
        memo = {}
        
        def dfs(idx, currSum):
            
            if (idx, currSum) in memo:
                return memo[(idx, currSum)]
            
            if currSum == 0:
                return True
            
            if idx == 0 or currSum < 0:
                return False
            memo[(idx, currSum)] = dfs(idx-1, currSum - nums[idx]) or dfs(idx-1, currSum)
            return memo[(idx, currSum)]
        
        return dfs(len(nums)-1, targetSum //2)
        

File: partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        
        int triangle_sum = 0; 
        
        for(int i : nums){
            triangle_sum += i;
        }
        
        if(triangle_sum % 2 != 0)return false;
        Map<String, Boolean> map = new HashMap<>();
        return dp(nums.length-1, triangle_sum/2, nums, map);
        
        
    }
    
    public static boolean dp(int idx, int target, int[] nums,Map<String, Boolean> map){
        String key = idx + " " + target;
        if(map.containsKey(key))return map.get(key);
        if(target == 0) return true;
        
        if(idx == 0 || target < 0) return false;
        
        map.put(key, dp(idx-1, target-nums[idx-1], nums,map) || dp(idx-1, target, nums,map));
        return map.get(key);
        
    }
}